addons/stock/models/product_strategy.py

```diff
@@ -80,13 +80,15 @@ def _domain_product_id(self):
     @api.onchange('location_in_id')
     def _onchange_location_in(self):
         child_location_count = 0
-        if self.location_out_id:
+        location_out = self.location_out_id._origin
+        location_in = self.location_in_id._origin
+        if location_out:
             child_location_count = self.env['stock.location'].search_count([
-                ('id', '=', self.location_out_id.id),
-                ('id', 'child_of', self.location_in_id.id),
-                ('id', '!=', self.location_in_id.id),
+                ('id', '=', location_out.id),
+                ('id', 'child_of', location_in.id),
+                ('id', '!=', location_in.id),
             ])
-        if not child_location_count or not self.location_out_id:
+        if not child_location_count or not location_out:
             self.location_out_id = self.location_in_id
```

Comment:
if both location in and out would be new id, this would be a little odd to replace the NewId out by False, but seems ok (and could be more hassle than worth to do all the case)

Reply:
@nle-odoo I think you are right thought, the domain is definitely not suitable because of the last clause  `!= false` when `location_in_id` does not have an origin. Since `location_in_id` is a required `Many2one` field I think it is save to just change the condition:
```diff
- if location_out:
+ if location_out and location_in:
```
Because `location_in` should only be false if there is no origin on the `location_in_id`.